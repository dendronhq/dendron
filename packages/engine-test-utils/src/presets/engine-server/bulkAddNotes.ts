import { NoteUtils } from "@dendronhq/common-all";
import { TestPresetEntryV4 } from "@dendronhq/common-test-utils";
import _ from "lodash";

const SCHEMAS = {};
const NOTES = {
  BASIC: new TestPresetEntryV4(async ({ vaults, engine }) => {
    const vault = vaults[0];
    const orig = _.size(engine.notes);
    const note1 = NoteUtils.create({
      id: "bar1",
      fname: "bar1",
      created: 1,
      updated: 1,
      vault,
    });
    const note2 = NoteUtils.create({
      id: "bar2",
      fname: "bar2",
      created: 1,
      updated: 1,
      vault,
    });
    await engine.bulkAddNotes({ notes: [note1, note2] });
    return [
      {
        actual: _.size(engine.notes),
        expected: orig + 2,
        msg: "should be 2 more notes",
      },
      {
        expected: engine.notes["bar1"],
        actual: note1,
      },
    ];
  }),
};
export const ENGINE_BULK_ADD_NOTES_PRESETS = {
  NOTES,
  SCHEMAS,
};
