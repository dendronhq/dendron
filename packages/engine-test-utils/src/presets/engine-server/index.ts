import { ENGINE_CONFIG_PRESETS } from "./config";
import { ENGINE_DELETE_PRESETS } from "./delete";
import { ENGINE_INFO_PRESETS } from "./info";
import { ENGINE_INIT_PRESETS } from "./init";
import { ENGINE_GET_NOTE_BLOCKS_PRESETS } from "./getNoteBlocks";
import NOTE_REF from "./note-refs";
import { ENGINE_QUERY_PRESETS } from "./query";
import { ENGINE_RENAME_PRESETS } from "./rename";
import { ENGINE_BULK_WRITE_NOTES_PRESETS } from "./bulkWriteNotes";
import { ENGINE_RENDER_PRESETS } from "./render";
import { ENGINE_WRITE_PRESETS, ENGINE_WRITE_PRESETS_MULTI } from "./write";
import _ from "lodash";
import { TestPresetEntryV4 } from "@dendronhq/common-test-utils";

export { ENGINE_HOOKS, ENGINE_HOOKS_BASE, ENGINE_HOOKS_MULTI } from "./utils";
export { ENGINE_RENAME_PRESETS };
export { ENGINE_QUERY_PRESETS };
export { ENGINE_WRITE_PRESETS };
export { ENGINE_CONFIG_PRESETS };

export const ENGINE_SERVER = {
  NOTE_REF,
  ENGINE_WRITE_PRESETS,
  ENGINE_INIT_PRESETS,
  ENGINE_DELETE_PRESETS,
  ENGINE_INFO_PRESETS,
  ENGINE_RENAME_PRESETS,
  ENGINE_GET_NOTE_BLOCKS_PRESETS,
  ENGINE_QUERY_PRESETS,
  ENGINE_BULK_WRITE_NOTES_PRESETS,
  ENGINE_RENDER_PRESETS,
};

type TestPresetEntry = TestPresetEntryV4;
type TestPresetDict = { [key: string]: TestPresetEntry };

/**
 * 
 @example
 *  test("", async () => {
 *    const TestCase= getPreset({presets: ENGINE_PRESETS, nodeType, presetName: "init", key: 'BAD_SCHEMA'})
 *    const { testFunc, ...opts } = TestCase;;
 *    await runEngineTestV5(testFunc, { ...opts, expect });
 *});
 * @param param0 
 * @returns 
 */
export const getPreset = ({
  presets,
  presetName,
  nodeType,
  key,
}: {
  presets: typeof ENGINE_PRESETS;
  presetName: string;
  nodeType: "SCHEMAS" | "NOTES";
  key: string;
}) => {
  const ent = _.find(presets, { name: presetName })!;
  // @ts-ignore
  const out = _.get(ent.presets[nodeType], key);
  if (!out) {
    throw Error(`no key ${key} found in ${presetName}`);
  }
  return out;
};

export const getPresetMulti = ({
  presets,
  presetName,
  nodeType,
  key,
}: {
  presets: typeof ENGINE_PRESETS_MULTI;
  presetName: string;
  nodeType: "NOTES";
  key: string;
}) => {
  const ent = _.find(presets, { name: presetName })!;
  // @ts-ignore
  const out = _.get(ent.presets[nodeType], key);
  if (!out) {
    throw Error(`no key ${key} found in ${presetName}`);
  }
  return out;
};

export const getPresetGroup = ({
  presets,
  presetName,
  nodeType,
}: {
  presets: typeof ENGINE_PRESETS;
  presetName: string;
  nodeType: "SCHEMAS" | "NOTES";
}) => {
  const ent = _.find(presets, { name: presetName })!;
  // @ts-ignore
  return ent.presets[nodeType] as TestPresetDict;
};

// ^iygzn9r2758w
export const ENGINE_PRESETS = [
  { name: "write", presets: ENGINE_SERVER.ENGINE_WRITE_PRESETS },
  {
    name: "bulkWriteNotes",
    presets: ENGINE_SERVER.ENGINE_BULK_WRITE_NOTES_PRESETS,
  },
  { name: "delete", presets: ENGINE_SERVER.ENGINE_DELETE_PRESETS },
  { name: "rename", presets: ENGINE_SERVER.ENGINE_RENAME_PRESETS },
  { name: "init", presets: ENGINE_SERVER.ENGINE_INIT_PRESETS },
  { name: "query", presets: ENGINE_SERVER.ENGINE_QUERY_PRESETS },
  { name: "info", presets: ENGINE_SERVER.ENGINE_INFO_PRESETS },
  {
    name: "getNoteBlocks",
    presets: ENGINE_SERVER.ENGINE_GET_NOTE_BLOCKS_PRESETS,
  },
  { name: "render", presets: ENGINE_SERVER.ENGINE_RENDER_PRESETS },
];

export const ENGINE_PRESETS_MULTI = [
  { name: "write", presets: ENGINE_WRITE_PRESETS_MULTI },
  { name: "delete", presets: ENGINE_SERVER.ENGINE_DELETE_PRESETS },
  { name: "rename", presets: ENGINE_SERVER.ENGINE_RENAME_PRESETS },
  { name: "init", presets: ENGINE_SERVER.ENGINE_INIT_PRESETS },
  { name: "query", presets: ENGINE_SERVER.ENGINE_QUERY_PRESETS },
];
