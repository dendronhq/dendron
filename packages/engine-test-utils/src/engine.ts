import {
  CleanDendronSiteConfig,
  CONSTANTS, DendronConfig, DEngineClient,
  DVault,
  DWorkspace,
  WorkspaceFolderRaw,
  WorkspaceOpts,
  WorkspaceSettings
} from "@dendronhq/common-all";
import {
  getDurationMilliseconds,
  tmpDir,
  vault2Path
} from "@dendronhq/common-server";
import {
  GenTestResults,
  PostSetupHookFunction,
  PreSetupHookFunction,
  RunEngineTestFunctionOpts,
  RunEngineTestFunctionV4,
  runJestHarnessV2,
  SetupHookFunction,
  SetupTestFunctionV4,
  sinon,
  TestResult
} from "@dendronhq/common-test-utils";
import { LaunchEngineServerCommand } from "@dendronhq/dendron-cli";
import {
  createEngine as engineServerCreateEngine,
  DConfig,
  WorkspaceService
} from "@dendronhq/engine-server";
import fs from "fs-extra";
import _ from "lodash";
import os from "os";
import path from "path";
import { SinonStub } from "sinon";
import { ENGINE_HOOKS } from "./presets";
import { GitTestUtils } from "./utils";

export type TestSetupWorkspaceOpts = {
  /**
   * Vaults to initialize engine with
   * Defaults to following if not set
   * [
   *    { fsPath: "vault1" },
   *    { fsPath: "vault2" },
   *    { fsPath: "vault3", name: "vaultThree" },
   *  ]
   */
  vaults?: DVault[];
  /**
   * Modify dendron config before initialization
   */
  modConfigCb?: (config: DendronConfig) => DendronConfig;
};

export type AsyncCreateEngineFunction = (
  opts: WorkspaceOpts
) => Promise<{engine: DEngineClient, port?: number}>;

/**
 * Create an {@link DendronEngine}
 */
export async function createEngineFromEngine(opts: WorkspaceOpts) {
  return {engine: engineServerCreateEngine(opts) as DEngineClient, port: undefined};
}

export { DEngineClient, DVault, WorkspaceOpts };

/**
 * Create a server
 * @param opts
 * @returns
 */
// @ts-ignore
export async function createServer(opts: WorkspaceOpts) {
  return await new LaunchEngineServerCommand().enrichArgs({
    wsRoot: opts.wsRoot,
  });
}

/**
 * Create an {@link DendronEngineClient}
 */
export async function createEngineFromServer(
  opts: WorkspaceOpts
): Promise<any> {
  const { engine, port } = await createServer(opts);
  await engine.init();
  return {engine, port};
}

export function createSiteConfig(
  opts: Partial<CleanDendronSiteConfig> &
    Required<Pick<CleanDendronSiteConfig, "siteRootDir" | "siteHierarchies">>
): CleanDendronSiteConfig {
  let copts = {
    siteNotesDir: "docs",
    siteUrl: "https://localhost:8080",
    ...opts,
  };
  return {
    ...copts,
    siteIndex: DConfig.getSiteIndex(copts),
  };
}

/**
 *
 * @param opts.asRemote: add git repo
 * @param opts.wsRoot: override given wsRoot
 * @returns
 */
export async function setupWS(opts: {
  vaults: DVault[];
  workspaces?: DWorkspace[];
  asRemote?: boolean;
  wsRoot?: string;
}) {
  const wsRoot = opts.wsRoot || tmpDir().name;
  const ws = new WorkspaceService({ wsRoot });
  ws.createConfig();
  const config = ws.config;
  let vaults = await Promise.all(
    opts.vaults.map(async (vault) => {
      await ws.createVault({ vault, config, updateConfig: false });
      return vault;
    })
  );
  config.vaults = _.sortBy(config.vaults, "fsPath");
  if (config.site.duplicateNoteBehavior) {
    config.site.duplicateNoteBehavior.payload = (
      config.site.duplicateNoteBehavior.payload as string[]
    ).sort();
  }
  ws.setConfig(config);
  if (opts.workspaces) {
    const vaultsFromWs = await _.reduce(
      opts.workspaces,
      async (resp, ent) => {
        await resp;
        await WorkspaceService.createWorkspace({
          wsRoot: path.join(wsRoot, ent.name),
          vaults: ent.vaults,
        });
        return ws.addWorkspace({ workspace: ent });
      },
      Promise.resolve({ vaults: [] } as { vaults: DVault[] })
    );
    vaults = vaults.concat(vaultsFromWs.vaults);
  }
  if (opts.asRemote) {
    await GitTestUtils.createRepoWithReadme(wsRoot);
  }
  return { wsRoot, vaults };
}

export type RunEngineTestV5Opts = {
  preSetupHook?: SetupHookFunction;
  createEngine?: AsyncCreateEngineFunction;
  extra?: any;
  expect: any;
  workspaces?: DWorkspace[];
  setupOnly?: boolean;
  initGit?: boolean;
  initHooks?: boolean;
  addVSWorkspace?: boolean;
  /**
   * Path to preset wsRoot
   */
  wsRoot?: string;
} & TestSetupWorkspaceOpts;

export type RunEngineTestFunctionV5<T = any> = (
  opts: RunEngineTestFunctionOpts & { extra?: any; engineInitDuration: number, port?: number }
) => Promise<TestResult[] | void | T>;

export class TestPresetEntryV5 {
  public preSetupHook: PreSetupHookFunction;
  public postSetupHook: PostSetupHookFunction;
  public testFunc: RunEngineTestFunctionV4;
  public extraOpts: any;
  public setupTest?: SetupTestFunctionV4;
  public genTestResults?: GenTestResults;
  public vaults: DVault[];
  public workspaces: DWorkspace[];

  constructor(
    func: RunEngineTestFunctionV5,
    opts?: {
      preSetupHook?: PreSetupHookFunction;
      postSetupHook?: PostSetupHookFunction;
      extraOpts?: any;
      setupTest?: SetupTestFunctionV4;
      genTestResults?: GenTestResults;
      vaults?: DVault[];
    }
  ) {
    let {
      preSetupHook,
      postSetupHook,
      extraOpts,
      setupTest,
      genTestResults,
      workspaces,
    } = _.defaults(opts, {
      workspaces: [],
    });
    this.preSetupHook = preSetupHook ? preSetupHook : async () => {};
    this.postSetupHook = postSetupHook ? postSetupHook : async () => {};
    this.testFunc = _.bind(func, this);
    this.extraOpts = extraOpts;
    this.setupTest = setupTest;
    this.genTestResults = _.bind(
      genTestResults ? genTestResults : async () => [],
      this
    );
    this.workspaces = workspaces;
    this.vaults = opts?.vaults || [
      { fsPath: "vault1" },
      { fsPath: "vault2" },
      {
        name: "vaultThree",
        fsPath: "vault3",
      },
    ];
  }
}

/**
 *
 * To create empty workspace, initilizae with `vaults = []`
 * @param func
 * @param opts.vaults: By default, initiate 3 vaults {vault1, vault2, (vault3, "vaultThree")}
 * @param opts.preSetupHook: By default, initiate empty
 * @param opts.wsRoot: Override the randomly generated test directory for the wsRoot
 * @returns
 */
export async function runEngineTestV5(
  func: RunEngineTestFunctionV5,
  opts: RunEngineTestV5Opts
): Promise<any> {
  const {
    preSetupHook,
    extra,
    vaults: vaultsInit,
    createEngine,
    initGit,
    workspaces,
    addVSWorkspace,
  } = _.defaults(opts, {
    preSetupHook: async ({}) => {},
    postSetupHook: async ({}) => {},
    createEngine: createEngineFromEngine,
    extra: {},
    // third vault has diff name
    vaults: [
      { fsPath: "vault1" },
      { fsPath: "vault2" },
      { fsPath: "vault3", name: "vaultThree" },
    ],
    addVSWorkspace: false,
  });
  try {
    // make sure tests don't overwrite local homedir contents
    TestEngineUtils.mockHomeDir();
    const { wsRoot, vaults } = await setupWS({
      vaults: vaultsInit,
      workspaces,
      wsRoot: opts.wsRoot,
    });
    if ((opts.initHooks, vaults)) {
      fs.ensureDirSync(path.join(wsRoot, CONSTANTS.DENDRON_HOOKS_BASE));
    }
    await preSetupHook({ wsRoot, vaults });
    const resp = await createEngine({ wsRoot, vaults });
    const engine = resp.engine;
    const start = process.hrtime();
    const initResp = await engine.init();
    if (addVSWorkspace) {
      fs.writeJSONSync(
        path.join(wsRoot, CONSTANTS.DENDRON_WS_NAME),
        {
          folders: vaults.map((ent) => ({
            path: ent.fsPath,
            name: ent.name,
          })) as WorkspaceFolderRaw[],
          settings: {},
          extensions: {},
        } as WorkspaceSettings,
        { spaces: 4 }
      );
    }
    const engineInitDuration = getDurationMilliseconds(start);
    const testOpts = {
      wsRoot,
      vaults,
      engine,
      initResp,
      port: resp.port,
      extra,
      config: engine,
      engineInitDuration,
    };
    if (initGit) {
      await GitTestUtils.createRepoForWorkspace(wsRoot);
      await Promise.all(
        vaults.map((vault) => {
          return GitTestUtils.createRepoWithReadme(
            vault2Path({ vault, wsRoot })
          );
        })
      );
    }
    if (opts.setupOnly) {
      return testOpts;
    }
    const results = (await func(testOpts)) || [];
    await runJestHarnessV2(results, expect);
    return { opts: testOpts, resp: undefined, wsRoot };
  } finally {
    // restore sinon so other tests can keep running
    sinon.restore();
  }
}

export function testWithEngine(
  prompt: string,
  func: RunEngineTestFunctionV4,
  opts?: Omit<RunEngineTestV5Opts, "expect"> & { only?: boolean }
) {
  if (opts?.only) {
    return test.only(prompt, async () => {
      await runEngineTestV5(func, {
        preSetupHook: ENGINE_HOOKS.setupBasic,
        ...opts,
        expect,
      });
    });
  } else {
    return test(prompt, async () => {
      await runEngineTestV5(func, {
        preSetupHook: ENGINE_HOOKS.setupBasic,
        ...opts,
        expect,
      });
    });
  }
}

export class TestEngineUtils {
  static mockHomeDir(dir?: string): SinonStub {
    if (_.isUndefined(dir)) dir = tmpDir().name;
    return sinon.stub(os, "homedir").returns(dir);
  }

  static vault1(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault1" })!;
  }

  static vault2(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault2" })!;
  }

  static vault3(vaults: DVault[]) {
    return _.find(vaults, { fsPath: "vault3" })!;
  }
}
