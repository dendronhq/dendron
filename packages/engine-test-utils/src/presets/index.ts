import { DVault, VaultUtils } from "@dendronhq/common-all";
import { NoteTestUtilsV4 } from "@dendronhq/common-test-utils";
import { WorkspaceOpts } from "../engine";

export * from "./engine-server";
export * from "./pods-core";
export * from "./vaults";

export enum SETUP_HOOK_KEYS {
  /**
   * alpha: link(beta)
   * beta: link(alpha)
   */
  WITH_LINKS = "WITH_LINKS",
}

const createLink = (fname: string, opts: { vaultPrefix?: DVault }) => {
  let cVaultPrefix = "";
  if (opts.vaultPrefix) {
    cVaultPrefix = VaultUtils.toURIPrefix(opts.vaultPrefix) + "/";
  }
  return `[[${cVaultPrefix}${fname}]]`;
};

export async function callSetupHook(
  key: SETUP_HOOK_KEYS,
  opts: {
    workspaceType: "single" | "multi";
    withVaultPrefix?: boolean;
  } & WorkspaceOpts
) {
  const { workspaceType, vaults, wsRoot, withVaultPrefix } = opts;
  const isMultiVault = workspaceType !== "single";
  const cVaults = isMultiVault ? vaults : [vaults[0], vaults[0]];

  // WITH LINKS
  if (key === SETUP_HOOK_KEYS.WITH_LINKS) {
    const link1 = createLink("beta", {
      vaultPrefix: withVaultPrefix ? cVaults[1] : undefined,
    });
    await NoteTestUtilsV4.createNote({
      fname: "alpha",
      wsRoot,
      vault: cVaults[0],
      body: link1,
    });
    const link2 = createLink("alpha", {
      vaultPrefix: withVaultPrefix ? cVaults[0] : undefined,
    });
    await NoteTestUtilsV4.createNote({
      fname: "beta",
      wsRoot,
      vault: cVaults[1],
      body: link2,
    });
    if (isMultiVault) {
      await NoteTestUtilsV4.createNote({
        fname: "beta",
        wsRoot,
        vault: cVaults[0],
        body: link2,
        genRandomId: true,
      });
    }
  } else {
    throw Error("not supported key");
  }
}
