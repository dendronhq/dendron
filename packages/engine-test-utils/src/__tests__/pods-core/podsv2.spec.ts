import { ResponseUtil, Time } from "@dendronhq/common-all";
import { tmpDir } from "@dendronhq/common-server";
import { NOTE_PRESETS_V4 } from "@dendronhq/common-test-utils";
import { EngineUtils, openPortFile } from "@dendronhq/engine-server";
import {
  AirtableExportPodV2,
  ConfigFileUtils,
  ExportPodConfigurationV2,
  ExternalConnectionManager,
  ExternalService,
  GoogleDocsExportPodV2,
  JSONExportPodV2,
  JSONSchemaType,
  MarkdownExportPodV2,
  NotionExportPodV2,
  PodExportScope,
  PodV2ConfigManager,
  PodV2Types,
  RunnableGoogleDocsV2PodConfig,
  RunnableNotionV2PodConfig,
  RunnableJSONV2PodConfig,
} from "@dendronhq/pods-core";
import fs from "fs-extra";
import _ from "lodash";
import path from "path";
import { runEngineTestV5 } from "../../engine";
import { ENGINE_HOOKS } from "../../presets";

/**
 * ConfigFileUtils
 */
describe("GIVEN a ConfigFileUtils class", () => {
  type testType = {
    foo: string;
    bar: number;
    nonEssential?: string;
  };

  const configSchema: JSONSchemaType<testType> = {
    required: ["foo", "bar"],
    type: "object",
    additionalProperties: true,
    properties: {
      foo: {
        description: "foo",
        type: "string",
      },
      bar: {
        description: "bar",
        type: "number",
      },
      nonEssential: {
        description: "optional",
        type: "string",
        nullable: true,
      },
    },
  };

  describe("WHEN generating a config file with a basic schema", () => {
    const fPath = path.join(tmpDir().name, "test_config.yml");

    afterAll(() => {
      fs.removeSync(fPath);
    });

    ConfigFileUtils.genConfigFileV2({
      fPath,
      configSchema,
    });

    test("THEN expect required properties to be present", () => {
      const readData = ConfigFileUtils.getConfigByFPath({ fPath });
      expect(readData).toBeDefined();
      expect(readData.foo).toEqual("TODO");
      expect(readData.bar).toEqual("TODO");
    });

    test("AND optional properties to be present, but commented out", () => {
      const configFile = fs.readFileSync(fPath, { encoding: "utf8" });
      expect(configFile).toMatchSnapshot();
    });
  });

  describe("WHEN there are pre-set properties", () => {
    const fPath = path.join(tmpDir().name, "test_config_with_presets.yml");

    afterAll(() => {
      fs.removeSync(fPath);
    });

    ConfigFileUtils.genConfigFileV2({
      fPath,
      configSchema,
      setProperties: {
        foo: "foo",
      },
    });

    test("THEN expect those properties to be set in the config", () => {
      const readData = ConfigFileUtils.getConfigByFPath({ fPath });
      expect(readData).toBeDefined();
      expect(readData.foo).toEqual("foo");
      expect(readData.bar).toEqual("TODO");
    });
  });
});

/**
 * PodV2ConfigManager
 */
describe("GIVEN a PodV2ConfigManager class", () => {
  const podsDir = tmpDir().name;

  afterAll(() => {
    fs.removeSync(podsDir);
  });

  const podConfig1: ExportPodConfigurationV2 = {
    podId: "foo",
    podType: PodV2Types.AirtableExportV2,
    exportScope: PodExportScope.Note,
  };

  const podConfig2: ExportPodConfigurationV2 = {
    podId: "bar",
    podType: PodV2Types.MarkdownExportV2,
    exportScope: PodExportScope.Hierarchy,
  };

  const podConfig3: ExportPodConfigurationV2 = {
    podId: "foo-bar",
    podType: PodV2Types.GoogleDocsExportV2,
    exportScope: PodExportScope.Note,
  };

  const podConfig4: ExportPodConfigurationV2 = {
    podId: "foo-bar",
    podType: PodV2Types.NotionExportV2,
    exportScope: PodExportScope.Note,
  };

  ConfigFileUtils.genConfigFileV2({
    fPath: path.join(podsDir, "foo.yml"),
    configSchema: AirtableExportPodV2.config(),
    setProperties: podConfig1,
  });

  ConfigFileUtils.genConfigFileV2({
    fPath: path.join(podsDir, "bar.yml"),
    configSchema: MarkdownExportPodV2.config(),
    setProperties: podConfig2,
  });

  ConfigFileUtils.genConfigFileV2({
    fPath: path.join(podsDir, "foo-bar.yml"),
    configSchema: GoogleDocsExportPodV2.config(),
    setProperties: podConfig3,
  });

  ConfigFileUtils.genConfigFileV2({
    fPath: path.join(podsDir, "notion.yml"),
    configSchema: NotionExportPodV2.config(),
    setProperties: podConfig4,
  });

  const podConfigWithoutExportScope = {
    podId: "test-config",
    podType: PodV2Types.MarkdownExportV2,
  };

  ConfigFileUtils.genConfigFileV2({
    fPath: path.join(podsDir, "test-config.yml"),
    configSchema: MarkdownExportPodV2.config(),
    setProperties: podConfigWithoutExportScope,
  });

  describe("WHEN getting a pod config by an existing ID", () => {
    test("THEN expect the pod config to be retrieved", () => {
      const podConfig = PodV2ConfigManager.getPodConfigById({
        podsDir,
        opts: {
          podId: "foo",
        },
      });
      expect(podConfig).toBeDefined();
      expect(podConfig?.podId).toEqual("foo");
      expect(podConfig?.podType).toEqual(PodV2Types.AirtableExportV2);
    });
  });

  describe("WHEN getting a pod config by a non-existent ID", () => {
    test("THEN expect nothing to be returned", () => {
      const podConfig = PodV2ConfigManager.getPodConfigById({
        podsDir,
        opts: {
          podId: "Invalid",
        },
      });

      expect(podConfig).toBeUndefined();
    });
  });

  describe("WHEN getting all pod configs", () => {
    test("THEN expect all pod configs to be returned", () => {
      const configs = PodV2ConfigManager.getAllPodConfigs(podsDir);

      expect(configs.length).toEqual(5);
    });
  });
});

/**
 * ExternalConnectionManager
 */
describe("GIVEN an ExternalConnectionManager class", () => {
  const testDir = tmpDir().name;
  const connectionManager = new ExternalConnectionManager(testDir);

  afterAll(() => {
    fs.removeSync(testDir);
  });

  afterEach(() => {
    fs.removeSync(path.join(testDir, ExternalConnectionManager.subPath));
  });

  describe("WHEN creating a new config with an existing ID", () => {
    test("THEN expect an error to be thrown", async () => {
      const fn = async () => {
        await connectionManager.createNewConfig({
          serviceType: ExternalService.Airtable,
          id: "foo",
        });

        await connectionManager.createNewConfig({
          serviceType: ExternalService.Airtable,
          id: "foo",
        });
      };

      expect(fn).rejects.toThrowError();
    });
  });

  describe("WHEN creating an AirtableConnection config with a unique ID", () => {
    test("THEN expect the config to be created", async () => {
      await connectionManager.createNewConfig({
        serviceType: ExternalService.Airtable,
        id: "foo",
      });

      const config = connectionManager.getConfigById({ id: "foo" });
      expect(config?.connectionId).toEqual("foo");
      expect(config?.serviceType).toEqual(ExternalService.Airtable);
    });
  });

  describe("WHEN creating a GoogleDocs config with a unique ID", () => {
    test("THEN expect the config to be created", async () => {
      await connectionManager.createNewConfig({
        serviceType: ExternalService.GoogleDocs,
        id: "foo",
      });

      const config = connectionManager.getConfigById({ id: "foo" });
      expect(config?.connectionId).toEqual("foo");
      expect(config?.serviceType).toEqual(ExternalService.GoogleDocs);
    });
  });

  describe("WHEN multiple configs of different types exist", () => {
    test("THEN expect type-matching configs to be returned", async () => {
      await connectionManager.createNewConfig({
        serviceType: ExternalService.Airtable,
        id: "airtable-one",
      });

      await connectionManager.createNewConfig({
        serviceType: ExternalService.Airtable,
        id: "airtable-two",
      });

      const configs = await connectionManager.getAllConfigsByType(
        ExternalService.Airtable
      );
      expect(configs.length).toEqual(2);
    });

    test("AND non type-matching configs to NOT be returned", async () => {
      await connectionManager.createNewConfig({
        serviceType: ExternalService.Airtable,
        id: "airtable",
      });

      await connectionManager.createNewConfig({
        serviceType: ExternalService.GoogleDocs,
        id: "googledocs",
      });

      const configs = await connectionManager.getAllConfigsByType(
        ExternalService.Airtable
      );
      expect(configs.length).toEqual(1);
    });
  });
});

/**
 * GoogleDocsExportPod
 */
describe("GIVEN a Google Docs Export Pod with a particular config", () => {
  // pod tests can take a long time to run
  jest.setTimeout(60000);

  describe("WHEN exporting a note", () => {
    test("THEN expect gdoc to be created", async () => {
      await runEngineTestV5(
        async (opts) => {
          const podConfig: RunnableGoogleDocsV2PodConfig = {
            exportScope: PodExportScope.Note,
            accessToken: "test",
            refreshToken: "test",
            expirationTime: Time.now().toSeconds() + 5000,
            connectionId: "foo",
          };
          const { wsRoot } = opts;
          const fpath = EngineUtils.getPortFilePathForCLI({ wsRoot });
          const port = openPortFile({ fpath });
          const pod = new GoogleDocsExportPodV2({
            podConfig,
            engine: opts.engine,
            port,
          });
          const response = {
            data: [
              {
                documentId: "testdoc",
                revisionId: "test",
                dendronId: "foo",
              },
            ],
            errors: [],
          };
          pod.createGdoc = jest.fn().mockResolvedValue(response);
          const props = (
            await opts.engine.findNotes({
              fname: "simple-wikilink",
              vault: opts.vaults[0],
            })
          )[0];

          const result = await pod.exportNotes([props]);
          const entCreate = result.data?.created!;
          const entUpdate = result.data?.updated!;
          expect(entCreate.length).toEqual(1);
          expect(entCreate[0]?.documentId).toEqual("testdoc");
          expect(entUpdate.length).toEqual(0);
        },
        {
          expect,
          preSetupHook: async ({ wsRoot, vaults }) => {
            await NOTE_PRESETS_V4.NOTE_WITH_WIKILINK_SIMPLE.create({
              wsRoot,
              vault: vaults[0],
            });
            await NOTE_PRESETS_V4.NOTE_WITH_WIKILINK_SIMPLE_TARGET.create({
              wsRoot,
              vault: vaults[0],
            });
            await fs.writeFileSync(
              path.join(wsRoot, ".dendron.port.cli"),
              "300"
            );
          },
        }
      );
    });
  });

  describe("WHEN there is an error in response", () => {
    test("THEN expect gdoc to return error message", async () => {
      await runEngineTestV5(
        async (opts) => {
          const podConfig: RunnableGoogleDocsV2PodConfig = {
            exportScope: PodExportScope.Note,
            accessToken: "test",
            refreshToken: "test",
            expirationTime: Time.now().toSeconds() + 5000,
            connectionId: "foo",
          };
          const { wsRoot } = opts;
          const fpath = EngineUtils.getPortFilePathForCLI({ wsRoot });
          const port = openPortFile({ fpath });
          const pod = new GoogleDocsExportPodV2({
            podConfig,
            engine: opts.engine,
            port,
          });
          const response = {
            data: [],
            errors: [
              {
                data: {},
                error: "error with status code 501",
              },
            ],
          };
          pod.createGdoc = jest.fn().mockResolvedValue(response);
          const props = (
            await opts.engine.findNotes({
              fname: "simple-wikilink",
              vault: opts.vaults[0],
            })
          )[0];

          const result = await pod.exportNotes([props]);
          const entCreate = result.data?.created!;
          expect(entCreate.length).toEqual(0);
          expect(ResponseUtil.hasError(result)).toBeTruthy();
        },
        {
          expect,
          preSetupHook: async ({ wsRoot, vaults }) => {
            await NOTE_PRESETS_V4.NOTE_WITH_WIKILINK_SIMPLE.create({
              wsRoot,
              vault: vaults[0],
            });
            await fs.writeFileSync(
              path.join(wsRoot, ".dendron.port.cli"),
              "300"
            );
          },
        }
      );
    });
  });
});

/**
 * Notion Export Pod
 */

describe("GIVEN a Notion Export Pod with a particular config", () => {
  describe("WHEN exporting a note", () => {
    test("THEN expect notion doc to be created", async () => {
      await runEngineTestV5(
        async (opts) => {
          const podConfig: RunnableNotionV2PodConfig = {
            exportScope: PodExportScope.Note,
            apiKey: "test",
            parentPageId: "test",
          };

          const pod = new NotionExportPodV2({
            podConfig,
          });
          const props = (
            await opts.engine.findNotes({
              fname: "bar",
              vault: opts.vaults[0],
            })
          )[0];
          const response = {
            data: [
              {
                dendronId: `${props.id}`,
                notionId: "test",
              },
            ],
            errors: [],
          };
          pod.convertMdToNotionBlock = jest.fn();
          pod.createPagesInNotion = jest.fn().mockResolvedValue(response);
          const result = await pod.exportNotes([props]);
          const entCreate = result.data?.created!;
          expect(entCreate.length).toEqual(1);
          expect(entCreate[0]?.notionId).toEqual("test");
        },
        {
          expect,
          preSetupHook: ENGINE_HOOKS.setupBasic,
        }
      );
    });
  });
});

/**
 * JSONExportPod
 */

describe("GIVEN a JSON Export Pod with a particular config", () => {
  describe("WHEN exporting a note and destination is clipboard", () => {
    test("THEN expect note to be exported", async () => {
      await runEngineTestV5(
        async (opts) => {
          const podConfig: RunnableJSONV2PodConfig = {
            exportScope: PodExportScope.Note,
            destination: "clipboard",
          };
          const pod = new JSONExportPodV2({
            podConfig,
          });
          const props = (
            await opts.engine.findNotes({
              fname: "bar",
              vault: opts.vaults[0],
            })
          )[0];
          const result = await pod.exportNotes([props]);
          const data = result.data?.exportedNotes!;
          expect(_.isString(data)).toBeTruthy();
          if (_.isString(data)) {
            expect(result.data?.exportedNotes).toEqual(
              JSON.stringify(props, null, 4)
            );
          }
        },
        {
          expect,
          preSetupHook: ENGINE_HOOKS.setupBasic,
        }
      );
    });
  });

  describe("When the destination is file system", () => {
    let exportDest: string;
    beforeAll(() => {
      exportDest = path.join(tmpDir().name, "export.json");
    });
    describe("WHEN exporting a note", () => {
      test("THEN expect note to be exported", async () => {
        await runEngineTestV5(
          async (opts) => {
            const podConfig: RunnableJSONV2PodConfig = {
              exportScope: PodExportScope.Note,
              destination: exportDest,
            };
            const pod = new JSONExportPodV2({
              podConfig,
            });

            const props = (
              await opts.engine.findNotes({
                fname: "bar",
                vault: opts.vaults[0],
              })
            )[0];

            await pod.exportNotes([props]);
            const content = fs.readFileSync(path.join(exportDest), {
              encoding: "utf8",
            });
            expect(content).toContain(JSON.stringify(props));
          },
          {
            expect,
            preSetupHook: ENGINE_HOOKS.setupBasic,
          }
        );
      });
    });
  });
});
