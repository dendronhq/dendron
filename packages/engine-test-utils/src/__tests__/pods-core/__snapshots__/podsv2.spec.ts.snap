// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GIVEN a ConfigFileUtils class WHEN generating a config file with a basic schema AND optional properties to be present, but commented out 1`] = `
"# description: foo
# type: string
# required: true
foo: TODO

# description: bar
# type: number
# required: true
bar: TODO

# description: optional
# type: string
# nonEssential: TODO"
`;
